test/test_pattern_matcher.py

```diff
@@ -50,14 +53,125 @@ def test_vin_permutations(self):
+    c1 = UOp(UOps.CONST, dtypes.int, arg=42)
+    a1 = UOp(UOps.CAST, dtypes.float, (c1,))
+    self.assert_equiv_uops(matcher.rewrite(a1), c0)
+
+  def test_zero_add(self):
+    matcher = PatternMatcher([({"uop": UOps.ALU, "arg": BinaryOps.ADD, "vin": [{"__name__": "x"},
+                                                                                      {"uop": UOps.CONST, "arg": 0}]}, lambda x: x)])
+    c0 = UOp(UOps.CONST, dtypes.float, arg=0.0)
+    x = UOp(UOps.CONST, dtypes.float, arg=55.0)
+    a1 = UOp(UOps.ALU, dtypes.float, (c0,x), BinaryOps.ADD)
+    a2 = UOp(UOps.ALU, dtypes.float, (x,c0), BinaryOps.ADD)
+    self.assertEqual(matcher.rewrite(a1), x)
+    self.assertEqual(matcher.rewrite(a2), x)
+
+  def test_sub_zero(self):
+    matcher = PatternMatcher([({"uop": UOps.ALU, "arg": BinaryOps.SUB, "vin": ({"__name__": "x"},
+                                                                                      {"uop": UOps.CONST, "arg": 0})}, lambda x: x)])
+    c0 = UOp(UOps.CONST, dtypes.float, arg=0.0)
+    x = UOp(UOps.CONST, dtypes.float, arg=42.0)
+    a1 = UOp(UOps.ALU, dtypes.float, (c0,x), BinaryOps.SUB)
+    a2 = UOp(UOps.ALU, dtypes.float, (x,c0), BinaryOps.SUB)
+    self.assertEqual(matcher.rewrite(a1), None)
+    self.assertEqual(matcher.rewrite(a2), x)
+
+  def test_zero_mul(self):
+    matcher = PatternMatcher([({"uop": UOps.ALU, "arg": BinaryOps.MUL, "vin": [{},
+                                                    {"__name__": "c", "uop": UOps.CONST, "arg": 0}]}, lambda c: c)])
+    c0 = UOp(UOps.CONST, dtypes.float, arg=0.0)
+    x = UOp(UOps.CONST, dtypes.float, arg=42.0)
+    a1 = UOp(UOps.ALU, dtypes.float, (c0,x), BinaryOps.MUL)
+    a2 = UOp(UOps.ALU, dtypes.float, (x,c0), BinaryOps.MUL)
+    self.assertEqual(matcher.rewrite(a1), c0)
+    self.assertEqual(matcher.rewrite(a2), c0)
+
+  def test_self_sub(self):
+    matcher = PatternMatcher([({"uop": UOps.ALU, "arg": BinaryOps.SUB, "vin": ({"__name__": "x"}, {"__name__": "x"})},
+                               lambda x: UOp.const(x.dtype, 0))])   # x-x -> 0
+    c0_int = UOp(UOps.CONST, dtypes.int, arg=0)
+    c0_float = UOp(UOps.CONST, dtypes.float, arg=0.0)
+    c1 = UOp(UOps.CONST, dtypes.int, arg=10)
+    c2 = UOp(UOps.CONST, dtypes.float, arg=55.55)
+    a1 = UOp(UOps.ALU, dtypes.int, (c1,c1), BinaryOps.SUB)
+    a2 = UOp(UOps.ALU, dtypes.float, (c2,c2), BinaryOps.SUB)
+    self.assert_equiv_uops(matcher.rewrite(a1), c0_int)
+    self.assert_equiv_uops(matcher.rewrite(a2), c0_float)
+
+  def test_fold_neg_mul(self):
+    matcher = PatternMatcher([({"uop": UOps.ALU, "arg": BinaryOps.MUL,
+                                       "vin": [{"__name__": "x"}, {"uop": UOps.CONST, "arg": -1}]}, lambda x: -x)])
+    c1 = UOp(UOps.CONST, dtypes.float, arg=42.0)
+    c2 = UOp(UOps.CONST, dtypes.float, arg=-1.0)
+    a1 = UOp(UOps.ALU, dtypes.float, (c1,c2), BinaryOps.MUL)
+    a2 = UOp(UOps.ALU, dtypes.float, (c1), UnaryOps.NEG)
```

Comment:
how does this pass the linter

Reply:
I fixed it for this test. Did some searching online and both ruff and pylint require custom rules to catch it.